File: java/05-Recursion/04-LinkedList-and-Recursion/src/Solution4.java
/**
 *  几乎和链表相关的所有操作, 都可以使用递归的形式完成
 *  建议对链表的增删改查, 进行递归实现
 *
 *  其他结构
 *      双端链表
 *      循环双端链表
 *      数组链表
 */
public class Solution4 {

    // 递归删除
    public ListNode removeElements(ListNode head, int val){
        // 结束条件
        if (head == null)
            return null;

        // 递归, 将当前节点(head)和其下一个节点(res)维持在一个时空, 是不是很神奇
//        ListNode res = removeElements(head.next,val);
//        if (head.val == val) {
//            return res;
//        } else {
//            head.next = res;
//            return head;
//        }

        head.next = removeElements(head.next,val);
        return (head.val == val) ? head.next : head;
    }

    public static void main(String[] args) {

        int [] nums = {1, 2, 6, 3, 4, 5, 6};
        ListNode head = new ListNode(nums);
        System.out.println(head);

        ListNode res = (new Solution4()).removeElements(head,6);
        System.out.println(res);
    }

}

File: go/05-Recursion/04-LinkedList-and-Recursion/solution4/solution4.go
package solution4

import (
	. "github.com/dingsw2019/play-with-data-struct/go/05-Recursion/04-LinkedList-and-Recursion/listnode"
)

/**
203 移除链表元素
https://leetcode-cn.com/problems/remove-linked-list-elements/

递归移除元素
*/
func removeElements(head *ListNode, val int) *ListNode {
	// 链表尾部
	if head == nil {
		return nil
	}
	// 定位到最后一个元素
	head.Next = removeElements(head.Next, val)
	// 从后往前开始处理, 跳过等于val的元素, 其余元素挂接到res
	if head.Val == val {
		return head.Next
	} else {
		return head
	}
}
